include/seqan3/utility/views/zip.hpp

```diff
@@ -6,26 +6,476 @@
+                {
+                    std::invoke(f, std::forward<ts>(elements))...
+                };
+        },
+        std::forward<tuple_t>(tuple));
+}
+
+template <typename fun_t, typename tuple_t>
+constexpr void tuple_for_each(fun_t && f, tuple_t && tuple)
+{
+    std::apply([&] <typename... ts> (ts && ... elements)
+        {
+            (std::invoke(f, std::forward<ts>(elements)), ...);
+        },
+        std::forward<tuple_t>(tuple));
+}
+
+template <bool is_const, typename... range_ts>
+concept all_random_access =
+    (std::ranges::random_access_range<view_helper::maybe_const<is_const, range_ts>> && ...);
+
+template <bool is_const, typename... range_ts>
+concept all_bidirectional =
+    (std::ranges::bidirectional_range<view_helper::maybe_const<is_const, range_ts>> && ...);
+
+template <bool is_const, typename... range_ts>
+concept all_forward =
+    (std::ranges::forward_range<view_helper::maybe_const<is_const, range_ts>> && ...);
+
+} // namespace seqan3::detail::zip
+
+template <std::ranges::input_range... Views>
+    requires (std::ranges::view<Views> && ...) && (sizeof...(Views) > 0)
+class zip_view : public std::ranges::view_interface<zip_view<Views...>>
+{
+    seqan3::common_tuple<Views...> views_;
+
+    template <bool>
+    class iterator;
+
+    template <bool>
+    class sentinel;
+
+public:
+    zip_view() requires (std::is_default_constructible_v<Views> && ...) = default;
+    constexpr explicit zip_view(Views... views) : views_(std::move(views)...)
+    {}
+
+    constexpr auto begin()
+        requires (!(view_helper::simple_view<Views> && ...))
+    {
+        return iterator<false>(zip::tuple_transform(std::ranges::begin, views_));
+    }
+
+    constexpr auto begin() const
+        requires (std::ranges::range<Views const> && ...)
+    {
+        return iterator<true>(zip::tuple_transform(std::ranges::begin, views_));
+    }
+
+    constexpr auto end()
+        requires (!(view_helper::simple_view<Views> && ...))
+    {
+        if constexpr (!zip::zip_is_common<Views...>)
+            return sentinel<false>(zip::tuple_transform(std::ranges::end, views_));
+        else if constexpr ((std::ranges::random_access_range<Views> && ...))
+            return begin() + std::iter_difference_t<iterator<false>>(size());
+        else
+            return iterator<false>(zip::tuple_transform(std::ranges::end, views_));
+    }
+
+    constexpr auto end() const
+        requires (std::ranges::range<Views const> && ...)
+    {
+        if constexpr (!zip::zip_is_common<Views const...>)
+            return sentinel<true>(zip::tuple_transform(std::ranges::end, views_));
+        else if constexpr ((std::ranges::random_access_range<Views const> && ...))
+            return begin() + std::iter_difference_t<iterator<true>>(size());
+        else
+            return iterator<true>(zip::tuple_transform(std::ranges::end, views_));
+    }
+
+    constexpr auto size()
+        requires (std::ranges::sized_range<Views> && ...)
```

Comment:
I think this has to be 'or', so either all Views are const (which should call `size() const`) or one of the Views is not const, so call this `size`.
```suggestion
        requires (std::ranges::sized_range<Views> || ...)
```

Also I am not sure if something similar goes for begin/end

Reply:
Yes, that makes sense!